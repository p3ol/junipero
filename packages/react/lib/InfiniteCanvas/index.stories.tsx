import { useRef, useState } from 'react';

import Button from '../Button';
import InfiniteCanvasZoom from '../InfiniteCanvasZoom';
import InfiniteCanvas, {
  type InfiniteCanvasCursorMode,
  type InfiniteCanvasBackgroundPattern,
  type InfiniteCanvasRef,
} from '.';

export default { title: 'react/InfiniteCanvas' };

export const Basic = () => {
  const canvasRef = useRef<InfiniteCanvasRef>(null);
  const [pattern, setPattern] = useState<
    InfiniteCanvasBackgroundPattern
  >('cross');
  const [cursorMode, setCursorMode] = useState<
    InfiniteCanvasCursorMode
  >('default');

  const [buttons, setButtons] = useState([]);

  return (
    <InfiniteCanvas
      ref={canvasRef}
      background={{
        pattern,
      }}
      cursorMode={cursorMode}
      className="fixed top-0 left-0 w-screen h-screen"
      onClick={() => {
        if (cursorMode !== 'add') {
          return;
        }

        setButtons(b => [
          ...b,
          canvasRef.current?.getCursorPosition(),
        ]);
      }}
      overlay={(
        <>
          <div className="absolute top-2 left-4 z-50 flex items-center gap-2">
            <Button
              onClick={() =>
                setPattern(
                  pattern === 'cross' ? 'dot' : 'cross',
                )
              }
            >
              Toggle Background Pattern ({pattern})
            </Button>
            <Button
              onClick={e => {
                e.stopPropagation();
                setCursorMode(
                  cursorMode === 'default'
                    ? 'pan' : cursorMode === 'pan' ? 'add' : 'default',
                );
              }}
            >
              Toggle Cursor Mode ({cursorMode})
            </Button>
          </div>
          <div
            className="flex items-center gap-2 absolute bottom-8 right-8 z-50"
          >
            <InfiniteCanvasZoom />
          </div>
        </>
      )}
    >
      <div
        className="w-full h-screen flex items-center justify-center"
      >
        <Button>Click me!</Button>
        { buttons.map((b, i) => (
          <Button key={i} style={{ position: 'absolute', left: b.x, top: b.y }}>
            Clicked!
          </Button>
        )) }
      </div>
    </InfiniteCanvas>
  );
};
