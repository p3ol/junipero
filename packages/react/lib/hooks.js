import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
