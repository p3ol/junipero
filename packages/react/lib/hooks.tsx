import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
