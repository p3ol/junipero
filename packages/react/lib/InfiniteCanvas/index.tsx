import {
  type ReactNode,
  type RefObject,
  useCallback,
  useEffect,
  useId,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { JuniperoRef, SpecialComponentPropsWithRef } from '../types';
import {
  type InfiniteCanvasContextType,
  InfiniteCanvasContext,
} from '../contexts';

export declare type InfiniteCanvasCursorMode =
  | 'default'
  | 'add'
  | 'pan';

export declare type InfiniteCanvasBackgroundPattern =
  | 'dot'
  | 'cross';

export declare interface InfiniteCanvasRef extends JuniperoRef {
  zoom: number;
  offsetX: number;
  offsetY: number;
  fitIntoView: (transitionDuration?: number) => void;
  setZoom: (newZoom: number, transitionDuration?: number) => void;
  zoomIn: (transitionDuration?: number) => void;
  zoomOut: (transitionDuration?: number) => void;
  getCursorPosition: () => { x: number; y: number };
  panTo: (x: number, y: number, transitionDuration?: number) => void;
  innerRef: RefObject<HTMLDivElement | null>;
  contentRef: RefObject<HTMLDivElement | null>;
  backgroundRef: RefObject<SVGSVGElement | null>;
}

export declare interface InfiniteCanvasProps extends
  SpecialComponentPropsWithRef<'div', InfiniteCanvasRef> {
  initialZoom?: number;
  minZoom?: number;
  maxZoom?: number;
  centerMargin?: number;
  cursorMode?: InfiniteCanvasCursorMode;
  background?: {
    pattern?: InfiniteCanvasBackgroundPattern;
    patternId?: string;
    gap?: number;
    size?: number;
    fill?: string;
    opacity?: number;
  };
  overlay?: ReactNode;
  globalEventsTarget?: EventTarget;
}

export declare interface InfiniteCanvasState {
  zoom: number;
  mouseX: number;
  mouseY: number;
  offsetX: number;
  offsetY: number;
  animate: number;
  panning: boolean;
  panStartX: number;
  panStartY: number;
}

const InfiniteCanvas = ({
  ref,
  children,
  overlay,
  className,
  background,
  initialZoom = 1,
  minZoom = 0.1,
  maxZoom = 10,
  centerMargin = 1000,
  cursorMode = 'default',
  globalEventsTarget = globalThis,
  ...rest
}: InfiniteCanvasProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const contentRef = useRef<HTMLDivElement>(null);
  const backgroundRef = useRef<SVGSVGElement>(null);
  const patternId = useId();
  const {
    gap = 20,
    size = 1,
    fill = 'var(--dots-color)',
    pattern = 'dot',
    patternId: customPatternId = patternId,
  } = background || {};
  const [state, dispatch] = useReducer(mockState<InfiniteCanvasState>, {
    zoom: initialZoom || 1,
    mouseX: 0,
    mouseY: 0,
    offsetX: 0,
    offsetY: 0,
    animate: 0,
    panning: false,
    panStartX: 0,
    panStartY: 0,
  });

  useTimeout(() => {
    dispatch({ animate: 0 });
  }, state.animate, [state.animate], { enabled: state.animate > 0 });

  useImperativeHandle(ref, () => ({
    zoom: state.zoom,
    offsetX: state.offsetX,
    offsetY: state.offsetY,
    mouseX: state.mouseX,
    mouseY: state.mouseY,
    fitIntoView,
    setZoom,
    zoomIn,
    zoomOut,
    getCursorPosition,
    panTo,
    innerRef,
    contentRef,
    backgroundRef,
    isJunipero: true,
  }));

  const onWheel = useCallback((e: WheelEvent) => {
    e.preventDefault();
    e.stopPropagation();

    // Zoom
    if (e.ctrlKey || e.metaKey) {
      const zoomFactor = Math.exp(-e.deltaY * 0.01);
      const newZoom = Math.max(minZoom,
        Math.min(maxZoom, state.zoom * zoomFactor));

      // Mouse position relative to canvas
      const rect = innerRef.current?.getBoundingClientRect();
      const mouseX = e.clientX - (rect?.left ?? 0);
      const mouseY = e.clientY - (rect?.top ?? 0);

      // Mouse position in content coordinates (before transform)
      const contentX = (mouseX - state.offsetX) / state.zoom;
      const contentY = (mouseY - state.offsetY) / state.zoom;

      // Adjust offset so zoom centers on mouse position
      const newOffsetX = mouseX - contentX * newZoom;
      const newOffsetY = mouseY - contentY * newZoom;

      dispatch({
        zoom: newZoom,
        offsetX: newOffsetX,
        offsetY: newOffsetY,
      });
    } else {
      // Pan only
      dispatch({
        offsetX: state.offsetX - e.deltaX,
        offsetY: state.offsetY - e.deltaY,
      });
    }
  }, [state.zoom, state.offsetX, state.offsetY, minZoom, maxZoom]);

  useEffect(() => {
    const ref = innerRef.current;
    ref?.addEventListener('wheel', onWheel, { passive: false });

    return () => {
      ref?.removeEventListener('wheel', onWheel);
    };
  }, [onWheel]);

  const fitIntoView = useCallback((transitionDuration?: number) => {
    if (!innerRef.current || !contentRef.current) {
      return;
    }

    const contentWidth = contentRef.current.scrollWidth + centerMargin * 2;
    const contentHeight = contentRef.current.scrollHeight + centerMargin * 2;

    const rect = innerRef.current.getBoundingClientRect();
    const canvasWidth = rect.width;
    const canvasHeight = rect.height;

    const zoomX = canvasWidth / contentWidth;
    const zoomY = canvasHeight / contentHeight;

    const newZoom = Math.max(Math.min(zoomX, zoomY, maxZoom), minZoom);

    // TODO: this is quite off due to margins but can't do better for now
    const newOffsetX = (canvasWidth - contentWidth * newZoom) / 2 +
      centerMargin / 2;
    const newOffsetY = (canvasHeight - contentHeight * newZoom) / 2 +
      centerMargin / 2;

    dispatch({
      zoom: newZoom,
      offsetX: newOffsetX,
      offsetY: newOffsetY,
      animate: transitionDuration ?? 100,
    });
  }, [minZoom, maxZoom, centerMargin]);

  const setZoom = useCallback((
    newZoom: number,
    transitionDuration?: number
  ) => {
    if (
      newZoom < minZoom ||
      newZoom > maxZoom ||
      !innerRef.current ||
      !contentRef.current
    ) {
      return;
    }

    const rect = innerRef.current.getBoundingClientRect();
    const canvasWidth = rect.width;
    const canvasHeight = rect.height;

    const contentWidth = contentRef.current.scrollWidth + centerMargin * 2;
    const contentHeight = contentRef.current.scrollHeight + centerMargin * 2;

    const newZoomClamped = Math.max(Math
      .min(newZoom, maxZoom), minZoom);
    const newOffsetX = (canvasWidth - contentWidth * newZoomClamped) / 2 +
      centerMargin / 2;
    const newOffsetY = (canvasHeight - contentHeight * newZoomClamped) / 2 +
      centerMargin / 2;

    dispatch({
      zoom: newZoomClamped,
      offsetX: newOffsetX,
      offsetY: newOffsetY,
      animate: transitionDuration ?? 100,
    });
  }, [minZoom, maxZoom, centerMargin]);

  const zoomIn = useCallback((transitionDuration?: number) => {
    const newZoom = (state.zoom || 1) * 1.2;
    setZoom(newZoom, transitionDuration);
  }, [state.zoom, setZoom]);

  const zoomOut = useCallback((transitionDuration?: number) => {
    const newZoom = (state.zoom || 1) / 1.2;
    setZoom(newZoom, transitionDuration);
  }, [state.zoom, setZoom]);

  const getCursorPosition = useCallback(() => {
    if (!innerRef.current) {
      return { x: 0, y: 0 };
    }

    return {
      x: Math.round((state.mouseX - state.offsetX) / state.zoom),
      y: Math.round((state.mouseY - state.offsetY) / state.zoom),
    };
  }, [state.mouseX, state.mouseY, state.zoom, state.offsetX, state.offsetY]);

  const panTo = useCallback((
    x: number,
    y: number,
    transitionDuration?: number
  ) => {
    if (!innerRef.current || !contentRef.current) {
      return;
    }

    const rect = innerRef.current.getBoundingClientRect();
    const canvasWidth = rect.width;
    const canvasHeight = rect.height;

    const newOffsetX = canvasWidth / 2 - x * state.zoom;
    const newOffsetY = canvasHeight / 2 - y * state.zoom;

    dispatch({
      offsetX: newOffsetX,
      offsetY: newOffsetY,
      animate: transitionDuration ?? 100,
    });
  }, [state.zoom]);

  const getContext = useCallback((): InfiniteCanvasContextType => ({
    zoom: state.zoom,
    offsetX: state.offsetX,
    offsetY: state.offsetY,
    mouseX: state.mouseX,
    mouseY: state.mouseY,
    fitIntoView,
    panTo,
    setZoom,
    zoomIn,
    zoomOut,
  }), [
    state.zoom, state.offsetX, state.offsetY, state.mouseX, state.mouseY,
    fitIntoView, setZoom, zoomIn, zoomOut, panTo,
  ]);

  const onMouseDown = useCallback((e: React.MouseEvent) => {
    if (cursorMode !== 'pan') {
      return;
    }

    e.preventDefault();
    e.stopPropagation();

    dispatch({
      panning: true,
      panStartX: e.clientX,
      panStartY: e.clientY,
    });
  }, [cursorMode]);

  const onMouseMove = useCallback((e: MouseEvent) => {
    if (!state.panning || cursorMode !== 'pan') {
      dispatch({ mouseX: e.clientX, mouseY: e.clientY });

      return;
    }

    const deltaX = e.clientX - state.panStartX;
    const deltaY = e.clientY - state.panStartY;

    dispatch({
      mouseX: e.clientX,
      mouseY: e.clientY,
      offsetX: state.offsetX + deltaX,
      offsetY: state.offsetY + deltaY,
      panStartX: e.clientX,
      panStartY: e.clientY,
    });
  }, [
    state.panStartX, state.panStartY, state.offsetX, state.offsetY,
    state.panning,
    cursorMode,
  ]);

  const onMouseUp = useCallback(() => {
    if (!state.panning || cursorMode !== 'pan') {
      return;
    }

    dispatch({
      panning: false,
      panStartX: 0,
      panStartY: 0,
    });
  }, [state.panning, cursorMode]);

  useEffect(() => {
    if (['pan', 'add'].includes(cursorMode)) {
      globalEventsTarget.addEventListener('mousemove', onMouseMove);
      globalEventsTarget.addEventListener('mouseup', onMouseUp);
    }

    return () => {
      globalEventsTarget.removeEventListener('mousemove', onMouseMove);
      globalEventsTarget.removeEventListener('mouseup', onMouseUp);
    };
  }, [
    onMouseMove, onMouseUp,
    globalEventsTarget, cursorMode,
  ]);

  const scaledPatternGap = gap * state.zoom;
  const scaledPatternSize = size * state.zoom;

  return (
    <InfiniteCanvasContext.Provider value={getContext()}>
      <div
        { ...rest }
        ref={innerRef}
        className={classNames(
          'junipero infinite-canvas',
          'cursor-mode-' + cursorMode,
          {
            'panning': state.panning,
          },
          className,
        )}
        onMouseDown={onMouseDown}
      >
        <div
          ref={contentRef}
          className={classNames(
            'infinite-canvas-content',
          )}
          style={{
            transform:
              `translate3d(${state.offsetX}px, ${state.offsetY}px, 0) ` +
              `scale3d(${state.zoom}, ${state.zoom}, 1)`,
            transformOrigin: '0 0',
            ...state.animate > 0 && {
              transition: `transform ${state.animate}ms ease-in-out`,
            },
          }}
        >
          { children }
        </div>

        { overlay }

        <svg ref={backgroundRef} className="infinite-canvas-background">
          <pattern
            id={customPatternId}
            patternUnits="userSpaceOnUse"
            x={state.offsetX % scaledPatternGap}
            y={state.offsetY % scaledPatternGap}
            width={scaledPatternGap}
            height={scaledPatternGap}
            patternTransform={
              `translate(${scaledPatternSize}, ${scaledPatternSize})`
            }
          >
            { pattern === 'cross' ? (
              <>
                <line
                  x1={scaledPatternSize * 2}
                  y1="0"
                  x2={scaledPatternSize * 2}
                  y2={scaledPatternSize * 4}
                  stroke={fill}
                  strokeWidth={scaledPatternSize / 4}
                />
                <line
                  x1="0"
                  y1={scaledPatternSize * 2}
                  x2={scaledPatternSize * 4}
                  y2={scaledPatternSize * 2}
                  stroke={fill}
                  strokeWidth={scaledPatternSize / 4}
                />
              </>
            ) : (
              <circle
                cx={scaledPatternSize}
                cy={scaledPatternSize}
                r={scaledPatternSize}
                fill={fill}
              />
            ) }
          </pattern>

          <rect
            x="0"
            y="0"
            width="100%"
            height="100%"
            fill={`url(#${customPatternId})`}
          />
        </svg>
      </div>
    </InfiniteCanvasContext.Provider>
  );
};

InfiniteCanvas.displayName = 'InfiniteCanvas';

export default InfiniteCanvas;
