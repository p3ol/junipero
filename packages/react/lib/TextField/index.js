import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
