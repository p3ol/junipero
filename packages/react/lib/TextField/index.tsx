import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
