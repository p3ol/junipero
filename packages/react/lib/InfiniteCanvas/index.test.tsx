import { render } from '@testing-library/react';

import InfiniteCanvas from '.';

describe('<InfiniteCanvas />', () => {
  it('should render an infinite canvas', () => {
    const { container, unmount } = render(
      <InfiniteCanvas>
        <div style={{ width: '1000px', height: '1000px' }}>Content</div>
      </InfiniteCanvas>
    );

    expect(container).toMatchSnapshot();
    unmount();
  });
});
