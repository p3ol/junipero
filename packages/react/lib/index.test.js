import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
