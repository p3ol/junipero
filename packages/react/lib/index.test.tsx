import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
