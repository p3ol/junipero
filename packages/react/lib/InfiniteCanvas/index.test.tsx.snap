// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<InfiniteCanvas /> should render an infinite canvas 1`] = `
<div>
  <div
    class="junipero infinite-canvas cursor-mode-default"
  >
    <div
      class="infinite-canvas-content"
      style="transform: translate3d(0px, 0px, 0) scale3d(1, 1, 1); transform-origin: 0 0;"
    >
      <div
        style="width: 1000px; height: 1000px;"
      >
        Content
      </div>
    </div>
    <svg
      class="infinite-canvas-background"
    >
      <pattern
        height="20"
        id="dots"
        patternTransform="translate(1, 1)"
        patternUnits="userSpaceOnUse"
        width="20"
        x="0"
        y="0"
      >
        <circle
          cx="1"
          cy="1"
          fill="var(--dots-color)"
          r="1"
        />
      </pattern>
      <rect
        fill="url(#dots)"
        height="100%"
        width="100%"
        x="0"
        y="0"
      />
    </svg>
  </div>
</div>
`;
